File: langs/typescript/src/viewmodels/critical-constraint-vm.ts
/**
 * CriticalConstraintVM — read-only wrapper for a CriticalConstraintM.
 *
 * Created by ScenarioVM after each solve run. No mutations; no commands.
 */
import { ComponentVMOf, NullMessageHub, NullDispatcher } from 'vmx';
import type { CriticalConstraintM } from '../models/critical-constraint.js';

export type CriticalConstraintVM = ComponentVMOf<CriticalConstraintM>;

export function makeCriticalConstraintVm(constraint: CriticalConstraintM): CriticalConstraintVM {
  const vm = ComponentVMOf.builder<CriticalConstraintM>()
    .name('criticalConstraint')
    .model(constraint)
    .services(NullMessageHub.INSTANCE, NullDispatcher.INSTANCE)
    .modeledHinter((m) => `idx=${m.constraintIndex} eliminated=${m.eliminated}`)
    .build();
  vm.construct();
  return vm;
}

File: langs/python/src/guidearch/viewmodels/critical_constraint_vm.py
"""CriticalConstraintVM — read-only wrapper for a CriticalConstraintM result.

Per spec/viewmodels.md §5.6: result VMs are read-only, no commands.
"""

from __future__ import annotations

from guidearch.models.critical_constraint import CriticalConstraintM


class CriticalConstraintVM:
    """Read-only view-model wrapper for a ranked critical constraint."""

    def __init__(self, model: CriticalConstraintM) -> None:
        self._model = model

    @property
    def constraint_index(self) -> int:
        return self._model.constraint_index

    @property
    def kind(self) -> str:
        return self._model.kind

    @property
    def eliminated(self) -> int:
        return self._model.eliminated

    @property
    def total(self) -> int:
        return self._model.total

    @property
    def redundant(self) -> bool:
        return self._model.redundant

    @property
    def model(self) -> CriticalConstraintM:
        return self._model
